File: safe-online-sdk-dotnet/src/main/linkid-sdk-dotnet/linkid-sdk-dotnet/src/main/cs/ws/attrib/AttributeClient.cs
/*
 * SafeOnline project.
 * 
 * Copyright 2006-2008 	Lin.k N.V. All rights reserved.
 * Lin.k N.V. proprietary/confidential. Use is subject to license terms.
 */
using System;
using System.Collections.Generic;

namespace safe_online_sdk_dotnet
{
	public interface AttributeClient
	{
		/// <summary>
		/// Gives back the attribute values of a single attribute of the given subject.
		/// </summary>
		/// <param name="userId"></param>
		/// <param name="attributeName"></param>
		/// <returns></returns>
		/// <exception cref="RuntimeException">Invalid response received</exception>
		/// <exception cref="AttributeNotFoundException">Attribute not found</exception>
		/// <exception cref="RequestDeniedException">Attribute Request denied</exception>
		/// <exception cref="AttributeUnavailableException">Attribute unavailable</exception>
        List<AttributeSDK> getAttributes(String userId, String attributeName);
		
		/// <summary>
		/// Gives back the attribute values via the specified dictionary of attributes. 
		/// The dictionary should hold the requested attribute names as keys. 
		/// The method will find the corresponding values.
		/// </summary>
		/// <param name="userId"></param>
        /// <param name="attributeMap"></param>
		/// <exception cref="RuntimeException">Invalid response received</exception>
		/// <exception cref="AttributeNotFoundException">Attribute not found</exception>
		/// <exception cref="RequestDeniedException">Attribute Request denied</exception>
		/// <exception cref="AttributeUnavailableException">Attribute unavailable</exception>
        void getAttributes(String userId, Dictionary<String, List<AttributeSDK>> attributeMap);
		
		/// <summary>
		/// Gives back a dictionary of attributes for the given subject that this application is allowed to read.
		/// </summary>
		/// <param name="userId"></param>
		/// <returns></returns>
		/// <exception cref="RuntimeException">Invalid response received</exception>
		/// <exception cref="AttributeNotFoundException">Attribute not found</exception>
		/// <exception cref="RequestDeniedException">Attribute Request denied</exception>
		/// <exception cref="AttributeUnavailableException">Attribute unavailable</exception>
		Dictionary<String, List<AttributeSDK>> getAttributes(String userId);
	}
}

File: linkid-sdk-api/src/main/java/net/link/safeonline/sdk/api/ws/attrib/client/AttributeClient.java
/*
 * linkID project.
 *
 * Copyright 2006-2014 Lin.k N.V. All rights reserved.
 * Lin.k N.V. proprietary/confidential. Use is subject to license terms.
 */

package net.link.safeonline.sdk.api.ws.attrib.client;

import java.io.Serializable;
import java.util.List;
import java.util.Map;
import net.link.safeonline.sdk.api.attribute.AttributeSDK;
import net.link.safeonline.sdk.api.exception.*;


/**
 * Interface for attribute client. Via components implementing this interface applications can retrieve attributes for subjects.
 * Applications can only retrieve attribute values for which the user confirmed the corresponding application identity.
 * <p/>
 * <p> The attribute value can be of type String, Boolean or an array of Object[] in case of a multivalued attribute. </p>
 *
 * @author fcorneli
 */
public interface AttributeClient {

    /**
     * Gives back attribute values via the map of attributes. The map should hold the requested attribute names as keys. The method will
     * fill in the corresponding values.
     */
    void getAttributes(String userId, Map<String, List<AttributeSDK<Serializable>>> attributes)
            throws AttributeNotFoundException, RequestDeniedException, WSClientTransportException, AttributeUnavailableException,
                   SubjectNotFoundException;

    /**
     * Gives back a map of attributes for the given subject that this application is allowed to read.
     */
    Map<String, List<AttributeSDK<Serializable>>> getAttributes(String userId)
            throws RequestDeniedException, WSClientTransportException, AttributeNotFoundException, AttributeUnavailableException,
                   SubjectNotFoundException;

    /**
     * Gives back a list of attributes for the given subject and attribute type that this application is allowed to read.
     */
    <T extends Serializable> List<AttributeSDK<T>> getAttributes(String userId, String attributeName)
            throws RequestDeniedException, WSClientTransportException, AttributeNotFoundException, AttributeUnavailableException,
                   SubjectNotFoundException;
}
